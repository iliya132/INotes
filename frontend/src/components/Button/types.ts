import { MouseEventHandler } from "react";

export interface IButtonProps {
    title: string,
    className?: string,
    disabled?: boolean;
    onClick?: MouseEventHandler<HTMLButtonElement>;
}