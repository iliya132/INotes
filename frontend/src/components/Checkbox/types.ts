/* eslint-disable no-unused-vars */
export interface ICheckboxProps {
    defaultValue?: boolean;
    id: string;
    className?: string;
    onChanged? : (newValue:boolean) => void;
    title?: string;
}
