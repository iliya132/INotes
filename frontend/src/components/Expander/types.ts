export interface IExpanderProps {
    isExpanded: boolean;
    expanderBody: JSX.Element;
    children: JSX.Element;
    onClick?: () => void;
}
