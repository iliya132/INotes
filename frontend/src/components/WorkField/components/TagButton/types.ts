import { INote } from "../../../../store/types";

export default interface ITagButtonProps {
    note: INote | null | undefined;
}

export interface ITag{
    label: string,
    value: string
}