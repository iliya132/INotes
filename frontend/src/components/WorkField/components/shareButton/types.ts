import { INote } from "../../../../store/types";

export default interface IShareBtnProps {
    note: INote | null | undefined;
}
