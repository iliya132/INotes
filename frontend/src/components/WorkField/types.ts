/* eslint-disable no-unused-vars */
import { INote } from "../../store/types";

export interface IWorkfieldProps {
    note?: INote;
}

export enum WfAction {
    Bold,
    Italic,
    Underscoped,
    Dotted,
    Number,
    H1,
    H2,
    H3,
    Code,
    Table
}
