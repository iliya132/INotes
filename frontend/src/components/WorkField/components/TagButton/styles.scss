@import '../../../../Misc/colors.scss';

.popup-container{
    min-width: 300px;
}

.selected-item{
    display: inline-block;
    background-color: $BlackCommon;
    border: 1px solid black;
    margin: 5px; 
    padding: 3px 10px;
    font-size: 14px;
    border-radius: 3px;
    color: white;
    user-select: none;

    &:hover {
        background-color: $BlackCommonHover;
        cursor: pointer;
        
    }
}