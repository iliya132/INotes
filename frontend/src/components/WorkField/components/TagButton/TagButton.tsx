import React, { useState } from "react";
import Popup from "reactjs-popup";
import SmallButton from "../../../SmallButton";
import { Icons } from "../../../Svg/types";
import ITagButtonProps, { ITag } from "./types";
import styles from './styles.scss';
import { OnChangeValue } from 'react-select';
import CreatableSelect from 'react-select/creatable';
import { INote } from "../../../../store/types";
import notesController from "../../../../controllers/NotesController";
import { useSelector } from "react-redux";
import { userTags } from "../../../../store/reducers/notebooksReducer";

export function TagButton(props: ITagButtonProps) {
    const { note } = props;
    if (!note) {
        return <></>;
    }
    const availableTags = useSelector(userTags)
    const selectedTags: ITag[] = note?.tags?.map(it => { return { label: it, value: it } })
    const options: ITag[] = availableTags?.map(it => { return { label: it, value: it } })

    const handleChange = (
        newValue: OnChangeValue<ITag, true>
    ) => {
        const val = newValue.map(it => it)
        const noteCopy: INote = { ...note, tags: val.map(it => it.value) };
        notesController.updateTags(noteCopy);
    };

    return (
        <Popup
            trigger={
                <div>
                    <SmallButton icon={Icons.Tag} tooltip="Задать тэг" data-testid="btn-tag" />
                </div>
            }
            disabled={note === undefined || note === null}
            position="bottom right">
            {(close) => {
                return (
                    <div className={styles["popup-container"]}>
                        <CreatableSelect
                            isMulti
                            hideSelectedOptions={false}
                            defaultValue={selectedTags}
                            onChange={handleChange}
                            options={options}
                        />
                    </div>
                )
            }}
        </Popup>
    )
}
