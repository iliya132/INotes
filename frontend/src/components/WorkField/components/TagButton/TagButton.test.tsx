/* eslint-disable no-undef */
import React from 'react';
import { createRoot, Root } from 'react-dom/client';
import { act } from 'react-dom/test-utils';
import { INote } from '../../../../store/types';
import { configureStore } from '@reduxjs/toolkit';
import { TagButton } from './TagButton';
import { RootState } from '../../../../store/store';
import { authSlice } from '../../../../store/reducers/authReducer';
import { notesSlice } from '../../../../store/reducers/notebooksReducer';
import { Provider } from 'react-redux';

//#region common setup
let root: Root | null = null;
let container: Element | null = null;

let initialState: RootState = {
    authReducer: {
        error: '',
        isAuth: true,
        user: { roles: ['user'], userName: 'test-user', avatarUrl: '/default.png' },
        validation: {
            isSucceded: true,
            errors: {},
        },
    },
    notebooksReducer: {
        notebooks: [],
        allNotes: [],
        userTags: [],
        selectedTags: []
    },
};

let store = configureStore({
    reducer: {
        authReducer: authSlice(initialState.authReducer).reducer,
        notebooksReducer: notesSlice(initialState.notebooksReducer).reducer,
    },
});

beforeEach(() => {
    act(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        root = createRoot(container);
    });
});

afterEach(() => {
    act(() => {
        root!.unmount();
        container!.remove();
        container = null;
    });
});
//#endregion

describe("test  input component", () => {
    it("matches snapshot", () => {
        const testNote:INote = {
            content: 'test',
            id: 0,
            isNew: false,
            isPublicUrlShared: false,
            name: 'test',
            parent: {color: 'red', id: 0, name: 'test'},
            publicUrl: 'url',
            tags: ['tag1', 'tag2']
        }
        act(() => {
            root!.render(<Provider store={store}><TagButton note={testNote} /></Provider>)
        })
        expect(container!.firstChild).toMatchSnapshot("Input test snapshot")
    })
})