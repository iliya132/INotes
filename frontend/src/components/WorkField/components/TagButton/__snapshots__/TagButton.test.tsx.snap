// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test  input component matches snapshot: Input test snapshot 1`] = `
<div
  aria-describedby="popup-1"
>
  <div
    class="small-button-container"
    data-testid="tag"
    title="Задать тэг"
  >
    <svg
      class="image small-button-icon tooltip"
      fill="BlackCommon"
      height="25"
      width="25"
    >
      <use
        xlink:href="test-file-stub#tag"
      />
    </svg>
  </div>
</div>
`;
